### PR TITLE
Fix checks related to value classes

### PR DESCRIPTION
**Pull request by liufengyun**
Fix checks related to value classes: #1670 #1642 

Note: despite the discussion today, the simplest fix is to move some `RefChecks` to `typer`. The TODO comment in the typer already points the direction for the migration:

```
+
+
+    // check value class constraints
+    RefChecks.checkDerivedValueClass(cls, body1)
+
     cdef1

     // todo later: check that
     //  1. If class is non-abstract, it is instantiatable:
     //  - self type is s supertype of own type
     //  - all type members have consistent bounds
     // 2. all private type members have consistent bounds
     // 3. Types do not override classes.
     // 4. Polymorphic type defs override nothing.

```

And I'd argue the complexity is manageable -- typer is not becoming more complex. And having checks earlier simplifies error handling and later processing.

Review @odersky .

### CODE REVIEW

File: src/dotty/tools/dotc/typer/RefChecks.scala [hunk 1]
```diff
@@ -720,6 +720,9 @@ object RefChecks {
           case List(param) =>
             if (param.is(Mutable))
               ctx.error("value class parameter must not be a var", param.pos)
+
+            if (isDerivedValueClass(param.info.typeSymbol))
+              ctx.error("value class may not wrap another user-defined value class", param.pos)
```
Comment (smarter) [inline src/dotty/tools/dotc/typer/RefChecks.scala]: No. That's completely fine and supported in Dotty. The only issue is when there is a cycle, this happens a value class underlying type (or the underlying type of its underlying type or ...) is itself. There is a `checkNonCyclic` method in `ExtensionMethods` that is supposed to do this but is currently disabled, try enabling it:
https://github.com/lampepfl/dotty/blob/master/src/dotty/tools/dotc/transform/ExtensionMethods.scala#L148-L150

Comment (liufengyun) [inline src/dotty/tools/dotc/typer/RefChecks.scala]: Thanks for the catch @smarter , I'm going to enable that.

File: src/dotty/tools/dotc/transform/TreeChecker.scala [hunk 2]
```diff
@@ -135,7 +135,7 @@ class TreeChecker extends Phase with SymTransformer {
     }
   }
 
-  class Checker(phasesToCheck: Seq[Phase]) extends ReTyper {
+  class Checker(phasesToCheck: Seq[Phase]) extends ReTyper with NoChecking {
```
Comment (odersky) [inline src/dotty/tools/dotc/transform/TreeChecker.scala]: That's a drastic change. Why disable all checking in TreeChecker?

If it's just checkDerivedValue class that should be disabled during tree checking then it's better to 
override that method directly.

**Comment by odersky** [comment]
Otherwise LGTM

File: src/dotty/tools/dotc/transform/TreeChecker.scala (hunk 2 above)
Comment (liufengyun) [inline src/dotty/tools/dotc/transform/TreeChecker.scala]: @odersky this is now addressed in the latest commit.

**Comment by odersky** [comment]
LGTM

**Pull request merged**